add case-insensitive mode

davibe opened the pull request:
Useful for matching/navigating xml document in a case-insensitive way (it does not affect parsing) #136

I had to add this for a project where server-side can't guarantee predictable casing. 
i.e.

```xml
<section>
  <ParaM kind="person" age="33" />
  <Param kind="persoN" age="34" />
</section>
```

I also thought there are more efficient ways to do this like lowercasing values in parse however it would mean to loose information which is particularly important in the case of attribute and element values.

drmohundro commented:
Hi @davibe - thanks for the PR!

I agree that lower-casing all values probably doesn't make sense. Maybe instead if there were a single method for looking so that the casing logic weren't duplicated would be better.

Last note - it looks like the CI build failed:

```
Compile Swift Module 'SWXMLHash' (2 sources)
/home/travis/build/drmohundro/SWXMLHash/Source/SWXMLHash.swift:293:55: error: missing argument for parameter 'caseInsensitive' in call
        let currentNode = parentStack.top().addElement(elementName, withAttributes: attributeNSDict)
                                                      ^
/home/travis/build/drmohundro/SWXMLHash/Source/SWXMLHash.swift:842:10: note: 'addElement(_:withAttributes:caseInsensitive:)' declared here
    func addElement(_ name: String, withAttributes attributes: NSDictionary, caseInsensitive:Bool) -> XMLElement {
         ^
/home/travis/build/drmohundro/SWXMLHash/Source/SWXMLHash.swift:374:55: error: missing argument for parameter 'caseInsensitive' in call
        let currentNode = parentStack.top().addElement(elementName, withAttributes: attributeNSDict)
                                                      ^
/home/travis/build/drmohundro/SWXMLHash/Source/SWXMLHash.swift:842:10: note: 'addElement(_:withAttributes:caseInsensitive:)' declared here
    func addElement(_ name: String, withAttributes attributes: NSDictionary, caseInsensitive:Bool) -> XMLElement {
```

davibe commented:
Strange, tomorrow i fix CI build

Where would you put such centralized casing logic? I am not sure how to make the conf visible in inner classes without passing it down

davibe commented:
Now the build failed for mysterious reasons that are not under my control

```
/Users/travis/.travis/job_stages: line 152: shell_session_update: command not found
```

davibe commented:
@drmohundro
- I rebased the patch on master removing conflicts.
- I have fixed errors in ci due to my code. One is still there but i don't think my patch is causing it. Right ? `/Users/travis/.travis/job_stages: line 152: shell_session_update: command not found`
- I have attempted to put the string case insensitive compare logic into one extension to kind of follow your suggestion. I still don't like the fact that i have to pass the `caseInsensitive` flag down like that.

drmohundro commented:
Thanks @davibe! And yeah, I agree with you that passing down the flag doesn't make much sense... if you don't mind, I'd like to pull down your branch and experiment with extracting it somehow... like maybe having a single "key lookup" method that can be overwritten.

The general thought process I've got (which totally might not work) is something like:

```
// PSEUDOCODE
node.lookupByKey(key)

// under config, maybe?
fun lookupByKey(key: String) {
  if config.lookupByKey != nil {
    // call the case insensitive matcher (or a different one)
  }
  else {
    // default lookup
  }
}
```

Like I said, it might not work, but it would configurable which might allow for a totally separate approach where someone might be able to plug in their own key lookup method entirely.

Also, regarding the TravicCI build, I'm honestly not sure what that is... I agree that it doesn't seem to be related to your changes, but I'll dig on that, too.

Thanks for your patience!

davibe commented:
@drmohundro how does `node` code access `config` ?
Are you suggesting to pass down the entire `config` object instead of just `caseInsensitive` flag ?

If you can suggest a way I can try something but otherwise I'd prefer to merge what we already have before new conflicts arise

drmohundro commented:
@davibe yeah, thanks for helping me think through this - I had a simpler picture in my head, but you're correct. Besides, any sort of refactoring or re-implementation would be behind the public interface so it should be fine. I just pushed a tiny commit to hopefully prevent Travis from failing. Once that builds, I'll merge and push out 4.1.0.

drmohundro commented:
Got the build passing! In case you were curious, the build error had nothing to do with shell_session_update, it was actually swiftlint - the errors were just further up in the build.

Your change is now live and tagged as `4.1.0`. Thanks again!

davibe commented:
Thank you @drmohundro !
